apps/scratch/src/pages/Solve.tsx

```diff
@@ -1,121 +1,77 @@
-            procedure: "getHeight",
-            result: document.body.scrollHeight,
-          });
-          break;
-        case "getSubmission":
-          if (vm) {
-            respondToMessageEvent(event, {
-              procedure: "getSubmission",
-              result: new Blob([vm.toJSON()], {
-                type: "application/json",
-              }),
+  const { hasLoaded, sendRequest } = useIframeParent(
+    useCallback(
+      async (request, respondToMessageEvent) => {
+        switch (request.procedure) {
+          case "getHeight":
+            respondToMessageEvent({
+              procedure: "getHeight",
+              result: document.body.scrollHeight,
             });
-          }
-          break;
-        case "getTask":
-          if (vm) {
-            saveCrtProject(vm).then((content) => {
-              respondToMessageEvent(event, {
-                procedure: "getTask",
-                result: content,
+            break;
+          case "getSubmission":
+            if (vm) {
+              respondToMessageEvent({
+                procedure: "getSubmission",
+                result: new Blob([vm.toJSON()], {
+                  type: "application/json",
+                }),
               });
-            });
-          }
-          break;
-        case "loadTask":
-          if (vm) {
-            const sb3Project = await message.arguments.arrayBuffer();
+            }
+            break;
+          case "getTask":
+            if (vm) {
+              saveCrtProject(vm).then((content) => {
+                respondToMessageEvent({
+                  procedure: "getTask",
+                  result: content,
+                });
+              });
+            }
+            break;
+          case "loadTask":
+            if (vm) {
+              const sb3Project = await request.arguments.arrayBuffer();
 
-            loadCrtProject(vm, sb3Project).then(() => {
-              respondToMessageEvent(event, {
-                procedure: "loadTask",
+              loadCrtProject(vm, sb3Project).then(() => {
+                respondToMessageEvent({
+                  procedure: "loadTask",
+                });
               });
-            });
-          }
-          break;
-        default:
-          break;
-      }
-    },
-    [vm],
+            }
+            break;
+          default:
+            break;
+        }
+      },
+      [vm],
+    ),
   );
 
-  // add an event listener to listen for messages from the iFrame
-  useEffect(() => {
-    if (isInIframe) {
-      window.addEventListener("message", handleParentWindowMessage);
-
-      return () => {
-        window.removeEventListener("message", handleParentWindowMessage);
-      };
-    }
-  }, [handleParentWindowMessage, isInIframe]);
+  const reportProgress = useCallback(
+    (assertions: number, passedAssertions: number) =>
+      sendRequest({
+        procedure: "reportProgress",
+        arguments: {
+          tests: assertions,
```

Comment:
I would strongly favor a `totalTests` name here, or `testsCount`, anything that points to this being a count/total/... 😅